src-monkey/Monkey.hs

```diff
@@ -0,0 +1,256 @@
+-- 100k elements, 1M bits (i.e. 10 bits per element),
+-- with a tiering setup with 3 levels and size multiplier of 4
+--
+-- >>> let t :: Num a => a; t = 4
+-- >>> let l :: Num a => a; l = 3
+--
+-- >>> let ex = monkeyBits 1_000_000 100_000 t l
+-- >>> ex
+-- [(1587,23721),(6349,76578),(25396,233034)]
+--
+-- We have almost 1.5 more memory used in first level compared to 'constantBits' allocation.
+--
+-- We can check that bits are divided between levels completely:
+--
+-- >>> sum $ map (\(_, bits) -> bits * (t - 1)) ex
+-- 999999
+--
+-- For the rest of computations we need bits and entries are floating numbers
+--
+-- >>> let ex' :: Floating a => [(a,a)]; ex' = map (bimap fromInteger fromInteger) ex
+--
+-- The false positive rates are constant across all levels:
+--
+-- >>> let ps = map (uncurry (falsePositiveRate @RDouble)) ex'
+-- >>> ps
+-- [0.00076,0.00304,0.01217]
+--
+-- compare that to to 'constantBits' @[0.00819,0.00819,0.00819]@ false-positive rates.
+-- First and second level are more precise, but the last one is less.
+--
+-- We can check again that the total memory is roughly the same we allocated initially
+--
+-- >>> totalMemory 100_000 t ps
+-- 984417.04432
+--
+-- Zero-result lookup cost, 5% (c.f. 7% with constant bits) of the time we'll do one I/O
+--
+-- >>> zeroResultCost t ps
+-- 0.04793
+--
+-- Non-zero lookup cost, 3.5% (c.f 6.5% with constant bits) of the time we'll do an extra I/O
+--
+-- >>> nonZeroResultCost t ps
+-- 1.03575
+--
+monkeyBits
+    :: Integer -- ^ total memory (in bits)
+    -> Integer -- ^ total elements
+    -> Integer -- ^ T: size ratio
+    -> Integer -- ^ L: level count
+    -> [(Integer, Integer) ]-- ^ elements and bits per run in a level
+monkeyBits m_max n t l
+    = zipWith (\(entries, _) bits -> (entries, round bits)) initial floating
+  where
+    initial  = constantBits  m_max n t l
+    initial_ = map (bimap fromInteger fromInteger) initial
+    floating = monkeyImpl (fromInteger m_max) (fromInteger t) initial_
+
+-- We optimise 'zeroResultCost' function given bits per each level,
+-- with a constraint that total memory stays the same
+--
+-- We enforce the constraint using substitution:
+-- bits_1 = total_memory - sum (bits_2 ... bits_L)
+--
+-- For now we use conjugate gradient descent instead of an analytical solution.
+-- Iterative version is easier to tweak.
+--
+monkeyImpl
+    :: Double
+    -> Double
+    -> [(Double,Double)]
+    -> [Double]
+monkeyImpl _ _ [] = []
+monkeyImpl m_max t ((entries0, _bits0) : initial) =
+    let rest = last $ take 100 $ conjugateGradientDescent target (map snd initial)
+    in (m_max / (t-1) - sum rest) : rest
+  where
+    target :: (Floating a, Mode a, Scalar a ~ Double) => [a] -> a
+    target ms = zeroResultCost (auto t) $ map (uncurry falsePositiveRate) $
+        (auto entries0 , auto (m_max / (t-1)) - sum ms) :
+        [ (auto entries, bits)
+        | ((entries, _), bits) <- zip initial ms
+        ]
```

Comment:
```
-- We enforce the constraint using substitution:
-- bits_1 = total_memory - sum (bits_2 ... bits_L)
```

Reply:
Yes, I saw that, but it isn't clear to me how that would not lead to a skewed distribution if we the first level's bits are used to ensure we don't go over budget

Reply:
I don't understand what you mean by skewed distribution.

Using substitution is the simplest method (when constraint is simple to be applicable) to enforce a constraint.
https://en.wikipedia.org/wiki/Constrained_optimization#Substitution_method

Let's maximise @x * y@, such that @x + y = 10@.

We can use substitution: @x = 10 - y@, so the function to optimise is @f (y) = (10 - y) * y = 10 * y - y * y@.
I.e. even there are two variables in the original problem statement, the coupling via equality constraint actually makes the problem to have only one variable.

Setting derivative to zero we can find that @y = 5@, and because @x = 10 - y = 10 - 5 = 5@, we get a result @x = 5; y = 5@.

Reply:
Okay, I understand now, thanks for the explanation